solr/core/src/java/org/apache/solr/cluster/maintenance/InactiveShardRemover.java

```diff
@@ -0,0 +1,251 @@
+
+  private final CoreContainer coreContainer;
+
+  private final DeleteActor deleteActor;
+
+  private ScheduledExecutorService executor;
+
+  private long scheduleIntervalSeconds;
+
+  private long ttlSeconds;
+
+  private int maxDeletesPerCycle;
+
+  /** Constructor invoked via Reflection */
+  public InactiveShardRemover(final CoreContainer cc) {
+    this(cc, new DeleteActor(cc));
+  }
+
+  @VisibleForTesting
+  InactiveShardRemover(final CoreContainer cc, final DeleteActor actor) {
+    this.coreContainer = cc;
+    this.deleteActor = actor;
+  }
+
+  @Override
+  public void configure(final InactiveShardRemoverConfig cfg) {
+    this.scheduleIntervalSeconds = cfg.scheduleIntervalSeconds;
+    this.maxDeletesPerCycle = cfg.maxDeletesPerCycle;
+    this.ttlSeconds = cfg.ttlSeconds;
+  }
+
+  @Override
+  public String getName() {
+    return PLUGIN_NAME;
+  }
+
+  @Override
+  public State getState() {
+    return state;
+  }
+
+  @Override
+  public void start() throws Exception {
+    state = State.STARTING;
+    executor = Executors.newScheduledThreadPool(1, new SolrNamedThreadFactory(PLUGIN_NAME));
+    executor.scheduleAtFixedRate(
+        this::deleteInactiveSlices,
+        scheduleIntervalSeconds,
+        scheduleIntervalSeconds,
+        TimeUnit.SECONDS);
+    state = State.RUNNING;
+  }
+
+  @Override
+  public void stop() {
+    if (state == State.RUNNING) {
+      state = State.STOPPING;
+      executor.shutdownNow();
+      try {
+        if (!executor.awaitTermination(10, TimeUnit.SECONDS)) {
+          log.warn(
+              "Executor pool did not terminate within the specified timeout: {} {}",
+              10,
+              TimeUnit.SECONDS);
+        }
+      } catch (InterruptedException e) {
+        log.warn("Failed to shut down the executor pool", e);
+        Thread.currentThread().interrupt();
+      }
+    }
+    state = State.STOPPED;
+  }
+
+  @VisibleForTesting
+  void deleteInactiveSlices() {
+    final ClusterState clusterState = coreContainer.getZkController().getClusterState();
+    Collection<Slice> inactiveSlices = new HashSet<>();
+    clusterState
+        .getCollectionsMap()
+        .forEach((k, v) -> inactiveSlices.addAll(collectInactiveSlices(v)));
+
+    if (log.isInfoEnabled()) {
+      log.info(
+          "Found {} inactive Shards to delete, {} will be deleted",
+          inactiveSlices.size(),
+          Math.max(inactiveSlices.size(), maxDeletesPerCycle));
+    }
+
+    inactiveSlices.stream().limit(maxDeletesPerCycle).forEach(this::deleteShard);
+  }
+
+  private Collection<Slice> collectInactiveSlices(final DocCollection docCollection) {
+    final Collection<Slice> slices = new HashSet<>(docCollection.getSlices());
+    slices.removeAll(docCollection.getActiveSlices());
+    return slices.stream().filter(this::isExpired).collect(Collectors.toList());
+  }
+
+  private void deleteShard(final Slice s) {
+    final String asyncId = s.getCollection() + ";" + s.getName() + ";DELETESHARD";
```

Comment:
Why use the "async" style?

Reply:
Honestly I was somewhat torn between the two. I chose async because I don't think the implementation really cares about waiting for the result of the delete operation, and a fire-and-forget approach seemed suitable. On the other hand, async makes testing a little more difficult. Happy to accept a recommendation or some guidance on the preferred style here.

Reply:
Since no process / user is waiting for the response, I think it's simpler to just call it synchronously.